Implement project-wide node groups

DarkMessiah opened the pull request:
This commit based on https://github.com/godotengine/godot/pull/24262.
Fix https://github.com/godotengine/godot-proposals/issues/3789 and maybe https://github.com/godotengine/godot-proposals/issues/3351 :)
*Bugsquad edit:*
- Fixes https://github.com/godotengine/godot-proposals/issues/907

* Global Groups - stored in the **godot.project** and visible to all nodes.
* Local Groups - these are just a list of groups derived from nodes in the current scene, excluding global groups.

**Global groups editor in Project Settings.**
<details><summary>Editor</summary>

![image](https://user-images.githubusercontent.com/1756388/168027683-ded5f8fd-a15c-4cba-9f34-c33bfa54f62e.png)
</details>

<details><summary>Remove global group dialog</summary>

![image](https://user-images.githubusercontent.com/1756388/168029290-0a13fcf0-5348-4e1a-9855-ba5eb95ea25b.png)
</details>

**Reworked Groups Editor.**
* Removed Manage Groups button + dialog
* Added filter for groups
* Added create button
* Reworked tree. Added categories for Local and Global groups.

<details><summary>Groups Editor</summary>

![image](https://user-images.githubusercontent.com/1756388/168029730-ce95adaf-8257-4718-bdbb-ba561d82255a.png)
</details>

<details><summary>Ability to create Global and Local groups.</summary>

![image](https://user-images.githubusercontent.com/1756388/168029769-0a220197-a3bd-421a-98fe-b5be55148be8.png)
</details>

<details><summary>Ability to Covert Global to Local and Local to Global. Rename. Delete.</summary>

![image](https://user-images.githubusercontent.com/1756388/168031864-d0685106-14b2-45e0-9221-541c8e9b858f.png)

![image](https://user-images.githubusercontent.com/1756388/168031908-5cf316f3-f405-48dc-aa38-1cf99e65af73.png)

![image](https://user-images.githubusercontent.com/1756388/168033837-19fa3656-153a-4962-a634-413ce93dfafb.png)

![image](https://user-images.githubusercontent.com/1756388/168033862-fac811ac-e9c4-492f-8708-aef2128ee854.png)
</details>

P.S. I have one problem. I store local groups cache inside `Map<Node *, Set<StringName>>`. When I opened scene, I'm storing all local groups into this Set where key is a `edited_scene_root_node`. I want to know the moment, when scene is closing and I can release a data from the Map.
I use `Map<Node *, Set<StringName>>` , because when you adding local group and unckeck it from all nodes, you can lose it when you are switching tabs.

DarkMessiah committed:
Implement project-wide node groups

Calinou commented:
Should "Local Groups" be renamed to "Scene Groups" to make it more obvious that those groups are scoped to the currently open scene in the editor?

fire-forge commented:
This is great! I like this workflow much better than the old one. 

A few suggestions:
* Move the + button to the right side. Most other Add/+ buttons throughout the editor are on the right.
![image](https://user-images.githubusercontent.com/67974470/168140363-1acacb05-a769-4d06-bb45-7c03c01f804d.png)
* Change the + button tooltip to just "Add a new group" instead of "Add/Create".
![image](https://user-images.githubusercontent.com/67974470/168140429-6a852aba-a735-484b-9b38-f61ba59a0e8f.png)
* Change the "Groups" tab to "Global Groups" in Project Settings to make it clear that this is only for global groups and not all groups.
![image](https://user-images.githubusercontent.com/67974470/168142733-4bd631a7-d591-4ad3-956d-e5efc6551be6.png)
* Split "Remove Completely" and "Make Local" into 2 buttons in the tree instead of combining them in the same button. The dialog is a bit confusing right now.
![image](https://user-images.githubusercontent.com/67974470/168143382-a318ab69-34e2-4544-a269-140f58733b52.png)

DarkMessiah commented:
What icon could I use for another remove button? :)

fire-forge commented on editor/groups_editor.cpp:
```diff
@@ -32,714 +32,679 @@
+void GroupsEditor::_show_remove_group_dialog() {
+	if (!remove_group_dialog) {
+		remove_group_dialog = memnew(ConfirmationDialog);
+		remove_group_dialog->set_title(TTR("Please confirm..."));
```
```suggestion
```

ConfirmationDialog title is "Please Confirm..." by default.

DarkMessiah replied:
Are you suggesting removing the title from the dialog? Delete dialogs that I saw have the title "Please confirm..." :)

fire-forge replied:
No, because `ConfirmationDialog.title` is "Please Confirm..." by default. So if this line is removed the title won't change.

![image](https://user-images.githubusercontent.com/67974470/169058417-f70dab05-1120-4039-ade2-a1746b299a3f.png)

fire-forge commented:
> What icon could I use for another remove button?

I have no idea. I looked around for a good icon to use and I can't find one. Maybe someone else has an idea?

DarkMessiah commented:
> I have no idea. I looked around for a good icon to use and I can't find one. Maybe someone else has an idea?

There is a Close icon, but I think it will be not clear for users.
![image](https://user-images.githubusercontent.com/1756388/168157028-721eeaf6-51ff-4399-827f-8fe55f933c81.png)

What do you think about CheckBox in dialog? :D
![image](https://user-images.githubusercontent.com/1756388/168157576-2beba3fc-8fac-4846-9fbd-ac7f54e0d3d2.png)

fire-forge commented:
A checkbox in the dialog is good. But you should make the text to the right of the checkbox be a part of the checkbox (using the `text` property) instead of a separate label, so that the whole thing can be clicked.

DarkMessiah commented:
* Rename Local groups to Scene groups
* Moved the + button to the right side and changed tooltip text.
![image](https://user-images.githubusercontent.com/1756388/168322164-76ad6ea1-64a7-44ef-a72c-cd15aa5cd365.png)

* Rename tab in Project Editor to Global Groups
* Redesigned removing dialog
![image](https://user-images.githubusercontent.com/1756388/168322085-c8e953f7-bed7-4847-beba-ce512ee99d68.png)

fire-forge commented:
It looks great! One change I would make is removing "This action is undoable!", because other actions that are undoable don't have anything saying that. Usually it's the other way around, with actions that can't be undone having a warning and everything else is just assumed to be undoable. 

I don't have any more suggestions regarding the design. I'll do a review of the code when I get a chance to :)

DarkMessiah commented:
Done!
<details><summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/1756388/168343686-47d80a09-1ce7-4b9d-8889-ec0a3760c7eb.png)

![image](https://user-images.githubusercontent.com/1756388/168343718-3060e1f3-d062-4af9-8b87-4f2b514fd674.png)

![image](https://user-images.githubusercontent.com/1756388/168343869-1f16c02d-ae03-4d72-ac89-76889b6c5ece.png)

![image](https://user-images.githubusercontent.com/1756388/168343971-4e86c3e5-6c21-4f22-944e-e177b4b6c8be.png)

</details>

fire-forge commented:
I would review this right now, but it doesn't compile :( The error messages are referring to HashMap, so I assume it's a difference introduced by https://github.com/godotengine/godot/pull/60881 or https://github.com/godotengine/godot/pull/60999.

https://github.com/godotengine/godot/runs/6427783069?check_suite_focus=true

DarkMessiah commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,498 @@
+}
+
+void GroupSettingsEditor::_item_edited() {
+	if (updating_group)
+		return;
```
Could you help with this? What I need to do? This is my first pull request with reviewing :)
1) Press "Commit suggestion"
2) Press "Resolve conversation"
3) Fix this in my branch and make force push with changes

fire-forge replied:
Sure! I find option 3 to be easiest, because pressing "Commit suggestion" makes a new commit which you then have to squash with the other commit. If you have to do that for several suggestions, then it makes a lot of commits to squash. 

"Resolve conversation" is for when the change is made, its decided that the change doesn't need to be made, or the change doesn't apply anymore. So here you can press it after you commit the suggestion.

DarkMessiah commented on editor/groups_editor.cpp:
```diff
@@ -32,714 +32,681 @@
+		item->set_meta("__local", true);
+		item->set_meta("__name", E->get());
+		item->set_meta("__description", "");
+		if (can_be_deleted) {
+			//item->add_button(0, get_theme_icon(SNAME("Remove"), SNAME("EditorIcons")), DELETE_GROUP);
+			item->add_button(0, get_theme_icon(SNAME("ActionCopy"), SNAME("EditorIcons")), COPY_GROUP);
```
Oh, I understood. It's for Instantiated Child Scenes. In original version, you cannot delete or copy groups from these nodes.
![image](https://user-images.githubusercontent.com/1756388/169314662-9cb8844f-f1cb-4b21-bd1a-9999dda5c496.png)

I added Lock icon for not editable groups. They aren't editable only for instantiated nodes.
 ![image](https://user-images.githubusercontent.com/1756388/169315169-12e02601-7702-4bd5-95a1-521cb93e962b.png)

I think what to do with Delete and Rename for these groups. They cannot be deleted and renamed, because they aren't part of current scene.

DarkMessiah commented:
@fire-forge I need to know when node added to the scene tree. Is there any signal for this?

DarkMessiah commented:
I removed my caching code, because it didn't work as I expected. Now the list of groups is updated every time when:
1) New node added
2) Any node removed
3) Scene tab changed

Also I added icons for groups from packed scenes:
![image](https://user-images.githubusercontent.com/1756388/177079346-2948a7bf-0be9-4342-9909-e5608806696e.png)

Groups from a packed scene cannot be removed or renamed outside of the packed scene.
![image](https://user-images.githubusercontent.com/1756388/177079504-08491d19-99ac-4faa-b0f3-5fbd0a57a39c.png)

When you'll try to remove these groups, you just remove them locally from the current scene, the packed scenes will not be changed. For the rename action I added TODO. Now I'm not showing anything, but maybe we need to show a warning message or create a new group.

DarkMessiah commented:
I think everything is ready for review :) 

![image](https://user-images.githubusercontent.com/1756388/182031436-34e4c59b-acda-4579-a36f-80652946f054.png)
1) You can't rename a group from a packed scene. (Nodes from packed scenes in scene tree have a similar behaviour.)
2) If you try to remove "blocked" group, you just remove it from other local nodes.

I took static method **can_edit** from this commit https://github.com/godotengine/godot/pull/62659, but I don't think that other features are necessary for a new way. You can uncheck a group and create a new one if you want.

DarkMessiah commented:
Rebased and resolved conflicts.

DarkMessiah commented:
- Fix some bugs after rebasing.
- Add caching for scene groups. Now you don't lose the list of scene groups after a scene tab change.

https://user-images.githubusercontent.com/1756388/197327444-17bf7bd2-b51c-4208-a950-43b5bbc7ed34.mp4

Mickeon commented:
I would like to bring this PR to attention in the next PR meeting, it would be nice to rebase it once more.

DarkMessiah commented:
@Mickeon Done!

mhilbrunner reviewed (commented):
PR review meeting:
We like the feature, but 4.0 development has entered feature freeze.
The groups also should be accessible easily from GDScript (@groud brought up looking at how autoloads are handled/exposed as an example to look at).

Suggestion by fales:
`ProjectSettings.get|add|remove_global_group(s)`

Thanks for contributing, sorry it took a while to get discussed in a PR meeting :)

Mickeon commented:
I just want to add, all contributors agree this is pretty much necessary. Being raw Strings, groups are really difficult to manage. For me, personally, this renders them essentially unusable, so this PR is very, very welcome even if it may not be ideal to merge it soon. Thank you very much for holding the torch.

MewPurPur commented:
What's the status of this?

DarkMessiah commented:
@MewPurPur I'll rebase it soon.

MewPurPur commented:
This button needs a tooltip:

![image](https://github.com/godotengine/godot/assets/85438892/d2a10e3d-ebb3-4e38-88aa-bc1b5cbe8e66)

Empty groups (e.g. only spaces) shouldn't be allowed in the global groups manager

AThousandShips commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,496 @@
+
+		undo_redo->commit_action();
+	} else if (tree->get_edited_column() == 1) {
+		// Description Edited
```
```suggestion
		// Description Edited.
```
The code style requires comments to start with a capital and end with a period, same everywhere

Calinou reviewed (changes requested):
Tested locally, it works as expected. Thanks for keeping it up with this PR :slightly_smiling_face: 

There are some UX issues I noticed however:

- The icon for creating a group should be moved to the left of the filter box. This is also consistent with the Scene tree dock. Otherwise, it looks like you're entering a group name to be added in the filter box:

![Screenshot_20230808_085242](https://github.com/godotengine/godot/assets/180032/0565eade-b371-48b0-8242-7d09867b5619)

![Screenshot_20230808_085237](https://github.com/godotengine/godot/assets/180032/c50e43e5-8f36-4e2c-a47c-a331e4da63e9)

___

- Pressing <kbd>Enter</kbd> in the create/rename group dialogs doesn't confirm the choice and close the dialog:

![Screenshot_20230808_085227](https://github.com/godotengine/godot/assets/180032/01238556-834d-41bc-b350-4110259d3b73)

___

- Double-clicking a group name doesn't toggle it. I'd expect it to do that, so that I don't have to aim for the checkbox precisely to toggle a group.
  - An alternative would be to rename when double-clicking, but renaming a group is a relatively uncommon operation compared to toggling a group.

![Screenshot_20230808_085156](https://github.com/godotengine/godot/assets/180032/edbb5deb-7ccc-4f3f-86fa-0a849d13daf4)

___

When removing a global group, "Delete references from all scenes." should **not** end with a period for consistency with other checkboxes of this style in the editor:

![image](https://github.com/godotengine/godot/assets/180032/b3735864-73f0-4443-8bc2-930a29457b1b)

YuriSizov commented:
By the way, for the validation panel in popups you should probably switch to using this new control: https://github.com/godotengine/godot/pull/78744

Also make sure to squash commits and amend the message. The current title of the PR should be a good option.

DarkMessiah commented:
Done!

@Calinou Now you can click on the text once to toggle it. But this behaviour seems very annoying. Maybe I'm used to the previous one :)

Calinou reviewed (approved):
All of my feedback was addressed. Great work!

Maran23 commented:
While testing, I found one minor UI 'issue'.
When a `CheckBox` is disabled, it has 4 more pixels of spacing for some reason:
![image](https://github.com/godotengine/godot/assets/66004280/61cd8e24-42ee-46f8-84a9-019a57c2413a)

Other than that, this looks very good and works very well.
Tested `Global Groups`, `Scene Groups`, delete (also without deleting all references), copy, revert with the `History` tab, inherited scenes and also `get_tree().get_nodes_in_group()` with the pretty nice auto completion :)

DarkMessiah commented:
`CheckBox` cannot be disabled in Tree. Therefore, I used the `GuiCheckedDisabled` icon and `TreeItem::CELL_MODE_STRING` for these elements.

Maran23 reviewed (approved):
Looks good. The minor UI issue can be fixed in a follow up PR.

AThousandShips commented on scene/main/node.cpp:
```diff
@@ -3011,6 +3011,11 @@ void Node::get_argument_options(const StringName &p_function, int p_idx, List<St
 	String pf = p_function;
 	if ((pf == "has_node" || pf == "get_node") && p_idx == 0) {
 		_add_nodes_to_options(this, this, r_options);
+	} else if ((pf == "add_to_group" || pf == "remove_from_group" || pf == "is_in_group") && p_idx == 0) {
```
```suggestion
	} else if (p_idx == 0 && (pf == "add_to_group" || pf == "remove_from_group" || pf == "is_in_group")) {
```
I'd move the cheaper check first

KoBeWi commented:
I found a few bugs when testing the functionality, but I can't reproduce them anymore :/

There is however a reproducible crash when in Node dock you delete global group and check to delete all references:
```
[0] Ref<SceneState>::ref (C:\godot_source\core\object\ref_counted.h:61)
[1] Ref<SceneState>::ref (C:\godot_source\core\object\ref_counted.h:61)
[2] Ref<SceneState>::Ref<SceneState> (C:\godot_source\core\object\ref_counted.h:167)
[3] PackedScene::get_state (C:\godot_source\scene\resources\packed_scene.cpp:1965)
[4] GroupSettingsEditor::remove_references (C:\godot_source\editor\group_settings_editor.cpp:233)
[5] call_with_variant_args_helper<GroupSettingsEditor,String const &,0> (C:\godot_source\core\variant\binder_common.h:303)
[6] call_with_variant_args_dv<GroupSettingsEditor,String const &> (C:\godot_source\core\variant\binder_common.h:451)
[7] MethodBindT<GroupSettingsEditor,String const &>::call (C:\godot_source\core\object\method_bind.h:333)
[8] Object::callp (C:\godot_source\core\object\object.cpp:741)
[9] Callable::callp (C:\godot_source\core\variant\callable.cpp:62)
[10] CallableCustomBind::call (C:\godot_source\core\variant\callable_bind.cpp:145)
[11] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[12] UndoRedo::_process_operation_list (C:\godot_source\core\object\undo_redo.cpp:333)
[13] UndoRedo::_redo (C:\godot_source\core\object\undo_redo.cpp:77)
[14] UndoRedo::commit_action (C:\godot_source\core\object\undo_redo.cpp:307)
[15] EditorUndoRedoManager::commit_action (C:\godot_source\editor\editor_undo_redo_manager.cpp:243)
[16] GroupsEditor::_confirm_delete (C:\godot_source\editor\groups_editor.cpp:606)
[17] call_with_variant_args_helper<GroupsEditor> (C:\godot_source\core\variant\binder_common.h:308)
[18] call_with_variant_args<GroupsEditor> (C:\godot_source\core\variant\binder_common.h:418)
[19] CallableCustomMethodPointer<GroupsEditor>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[20] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[21] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1073)
[22] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3586)
[23] Object::emit_signal<> (C:\godot_source\core\object\object.h:891)
[24] AcceptDialog::_ok_pressed (C:\godot_source\scene\gui\dialogs.cpp:127)
[25] call_with_variant_args_helper<AcceptDialog> (C:\godot_source\core\variant\binder_common.h:308)
[26] call_with_variant_args<AcceptDialog> (C:\godot_source\core\variant\binder_common.h:418)
[27] CallableCustomMethodPointer<AcceptDialog>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[28] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[29] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1073)
[30] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3586)
[31] Object::emit_signal<> (C:\godot_source\core\object\object.h:891)
[32] BaseButton::_pressed (C:\godot_source\scene\gui\base_button.cpp:139)
[33] BaseButton::on_action_event (C:\godot_source\scene\gui\base_button.cpp:177)
[34] BaseButton::gui_input (C:\godot_source\scene\gui\base_button.cpp:70)
[35] Control::_call_gui_input (C:\godot_source\scene\gui\control.cpp:1806)
[36] Viewport::_gui_call_input (C:\godot_source\scene\main\viewport.cpp:1586)
[37] Viewport::_gui_input_event (C:\godot_source\scene\main\viewport.cpp:1852)
[38] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:3163)
[39] Window::_window_input (C:\godot_source\scene\main\window.cpp:1526)
[40] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:303)
[41] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:418)
[42] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[43] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[44] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:2619)
[45] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:2582)
[46] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:723)
[47] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:981)
[48] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:2303)
[49] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1479)
[50] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:182)
[51] _main (C:\godot_source\platform\windows\godot_windows.cpp:204)
[52] main (C:\godot_source\platform\windows\godot_windows.cpp:218)
[53] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:232)
[54] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[55] <couldn't map PC to fn name>
```

KoBeWi commented on core/config/project_settings.cpp:
```diff
@@ -276,6 +276,11 @@ bool ProjectSettings::_set(const StringName &p_name, const Variant &p_value) {
 				remove_autoload(node_name);
 			}
+		} else if (p_name.operator String().begins_with("global_group/")) {
+			String node_name = p_name.operator String().split("/")[1];
```
```suggestion
			String group_name = p_name.operator String().get_slice("/", 1);
```

KoBeWi commented on core/config/project_settings.cpp:
```diff
@@ -321,6 +326,9 @@ bool ProjectSettings::_set(const StringName &p_name, const Variant &p_value) {
 			}
 			add_autoload(autoload);
+		} else if (p_name.operator String().begins_with("global_group/")) {
+			String group_name = p_name.operator String().split("/")[1];
```
```suggestion
			String group_name = p_name.operator String().get_slice("/", 1);
```
The autoload code does this wrong. `get_slice()` is better here.

KoBeWi commented on core/config/project_settings.h:
```diff
@@ -104,6 +104,8 @@ class ProjectSettings : public Object {
 
 	HashMap<StringName, AutoloadInfo> autoloads;
 
+	HashMap<StringName, String> global_groups;
```
```suggestion
	HashMap<StringName, String> global_groups;
```
I think the separation is unnecessary here.

KoBeWi commented on core/config/project_settings.cpp:
```diff
@@ -1205,6 +1213,24 @@ ProjectSettings::AutoloadInfo ProjectSettings::get_autoload(const StringName &p_
+}
+
+void ProjectSettings::add_global_group(const StringName &p_name, const String &p_description) {
+	ERR_FAIL_COND_MSG(p_name == StringName(), "Trying to add group with no name.");
```
```suggestion
	ERR_FAIL_COND_MSG(p_name == StringName(), "Trying to add global group with no name.");
```

KoBeWi commented on core/config/project_settings.cpp:
```diff
@@ -1205,6 +1213,24 @@ ProjectSettings::AutoloadInfo ProjectSettings::get_autoload(const StringName &p_
+}
+
+void ProjectSettings::remove_global_group(const StringName &p_name) {
+	ERR_FAIL_COND_MSG(!global_groups.has(p_name), "Trying to remove non-existent group.");
```
```suggestion
	ERR_FAIL_COND_MSG(!global_groups.has(p_name), "Trying to remove non-existent global group.");
```

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+			item->set_icon(0, get_theme_icon(SNAME("GuiCheckedDisabled"), SNAME("EditorIcons")));
+		}
+		if (!scene_groups[E]) {
+			item->add_button(0, get_theme_icon(SNAME("Lock"), SNAME("EditorIcons")), -1, true, TTR("This group can't be edited outside a packed scene"));
```
```suggestion
			item->add_button(0, get_theme_icon(SNAME("Lock"), SNAME("EditorIcons")), -1, true, TTR("This group belongs to another scene and can't be edited."));
```
This option shows up in inherited scenes too.

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,450 @@
+		undo_redo->add_do_property(ProjectSettings::get_singleton(), old_property_name, Variant());
+		undo_redo->add_undo_property(ProjectSettings::get_singleton(), old_property_name, description);
+
+		undo_redo->add_do_method(this, "rename_references", old_name, new_name);
+		undo_redo->add_undo_method(this, "rename_references", new_name, old_name);
```
A method that scans every scene is part of undoredo action 😬
It's nice that you can undo it, but going through all scenes can take many minutes in a big project. I don't know how to handle it better though, other than making a separate refactoring tool. I guess it's something for another PR.

KoBeWi commented on editor/group_settings_editor.h:
```diff
@@ -0,0 +1,99 @@
+	GDCLASS(GroupSettingsEditor, VBoxContainer);
+
+	const String GLOBAL_GROUP_PREFIX = "global_group/";
+	String group_changed = "group_changed";
```
```suggestion
	StringName group_changed = "group_changed";
```
`emit_signal()` operates takes StringName.

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,450 @@
+	if (!ti) {
+		return;
+	}
+
+	ti->select(0);
+
+	_show_remove_dialog();
```
```suggestion
	ti->select(0);
	_show_remove_dialog();
```
Too many blanks :/
You don't need to separate every single line.

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,450 @@
+
+void GroupSettingsEditor::_get_all_scenes(EditorFileSystemDirectory *p_dir, RBSet<String> &r_list) {
+	for (int i = 0; i < p_dir->get_file_count(); i++) {
+		if (p_dir->get_file_type(i) == "PackedScene")
+			r_list.insert(p_dir->get_file_path(i));
```
```suggestion
		if (p_dir->get_file_type(i) == "PackedScene") {
			r_list.insert(p_dir->get_file_path(i));
		}
```

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,450 @@
+
+	for (const StringName &E : keys) {
+		TreeItem *item = tree->create_item(root);
+		item->set_meta("__name", E);
+		item->set_meta("__description", groups_cache[E]);
```
What's the purpose of this meta? Isn't the text enough?

DarkMessiah replied:
I use meta in `_item_edited` to compare the old name/description with the new one.

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,450 @@
+	_add_group(group_name->get_text(), group_description->get_text());
+}
+
+void GroupSettingsEditor::_changed() {
```
Is this method used anywhere?

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,450 @@
+	}
+
+	remove_check_box->set_pressed(false);
+	remove_label->set_text(vformat(TTR("Delete group \"%s\" and its description?"), ti->get_text(0)));
```
```suggestion
	remove_label->set_text(vformat(TTR("Delete group \"%s\"?"), ti->get_text(0)));
```
No need to ask about description; it's part of the group (somewhat).

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,450 @@
+	tree->set_columns(3);
+	tree->set_column_titles_visible(true);
+
+	tree->set_column_title(0, TTR("Name"));
+	tree->set_column_expand(0, true);
+
+	tree->set_column_title(1, TTR("Description"));
+	tree->set_column_expand(1, true);
+
+	tree->set_column_expand(2, false);
```
```suggestion
	tree->set_column_title(0, TTR("Name"));
	tree->set_column_title(1, TTR("Description"));
	tree->set_column_expand(2, false);
```
Columns are expanded by default.

KoBeWi commented on editor/group_settings_editor.h:
```diff
@@ -0,0 +1,99 @@
+#ifndef GROUP_SETTINGS_EDITOR_H
+#define GROUP_SETTINGS_EDITOR_H
+
+#include "core/templates/hash_map.h"
```
I think the hashmap include can be removed.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
-	if (p_current != scene_tree->get_edited_scene_root()) {
-		item_name = String(p_current->get_parent()->get_name()) + "/" + item_name;
+void GroupsEditor::_remove_scene_group(const String &p_name) {
+	if (!node) {
```
Why is a node required here?

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
-
-	if (!selected) {
-		return;
+void GroupsEditor::_rename_node_references(Node *p_node, const String &p_old_name, const String &p_new_name) {
```
I think you could make the GroupSettingsEditor method public and call it here instead of duplicating code.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+	}
 
-	undo_redo->commit_action();
+	switch (p_id) {
```
switch for single option is redundant.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+	filter->set_clear_button_enabled(true);
+	filter->set_placeholder(TTR("Filter Groups"));
+	filter->set_h_size_flags(Control::SIZE_EXPAND_FILL);
+	filter->connect("text_changed", callable_mp(this, &GroupsEditor::_filter_changed));
```
```suggestion
	filter->connect("text_changed", callable_mp(this, &GroupsEditor::_update_tree).unbind(1));
```
`_filter_changed()` can be removed.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
 
-	String name = p_name.strip_edges();
-	if (name.is_empty() || groups->get_item_with_text(name)) {
+void GroupsEditor::_update_groups() {
```
Unless there is a reason to update the groups while the editor is invisible, this function can be optimized:
```C++
if (!is_visible_in_tree()) {
	groups_dirty = true;
	return;
}
groups_dirty = false
```
And then in _notification:
```C++
case NOTIFICATION_VISIBILITY_CHANGED: {
	if (groups_dirty && is_visible_in_tree()) {
		_update_groups();
	}
```
This is important, because it affects the whole editor. There is no need to update group list if the user is not currently using it.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+	updating_groups = false;
+}
+
+void GroupsEditor::_update_tree() {
```
This can also be optimized as I described above.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+void GroupsEditor::_node_removed(Node *p_node) {
+	if (scene_root_node == p_node) {
+		scene_groups_for_caching = scene_groups;
+		call_deferred(SNAME("_cache_scene_groups"), p_node);
```
```suggestion
		callable_mp(this, &GroupsEditor::_cache_scene_groups).call_deferred(p_node);
```

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
-	node = p_node;
-	update_tree();
-}
+void GroupsEditor::_bind_methods() {
```
This method and all binds can be removed and replaced with `callable_mp()`. I gave an example in `_node_removed()`.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
-			add_filter->set_clear_button_enabled(true);
-			remove_filter->set_right_icon(groups->get_theme_icon(SNAME("Search"), SNAME("EditorIcons")));
-			remove_filter->set_clear_button_enabled(true);
+		case NOTIFICATION_READY: {
```
```suggestion
		case NOTIFICATION_THEME_CHANGED: {
```
Also SNAME can be removed.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
 	}
 
+	if (p_mouse_button == MouseButton::LEFT) {
+		call_deferred("_item_edited");
```
Can also become `callable_mp()`.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+		vbc->add_child(gc);
+
+		Label *label_name = memnew(Label(TTR("Name:")));
+		label_name->set_h_size_flags(0); // Size flags: shrink begin.
```
```suggestion
		label_name->set_h_size_flags(SIZE_SHRINK_BEGIN);
```
Same with others.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+		if (!scene_groups[E]) {
+			item->add_button(0, get_theme_icon(SNAME("Lock"), SNAME("EditorIcons")), -1, true, TTR("This group can't be edited outside a packed scene"));
+		}
+		item->add_button(0, get_theme_icon(SNAME("ActionCopy"), SNAME("EditorIcons")), COPY_GROUP, false, TTR("Copy group to clipboard"));
```
```suggestion
		item->add_button(0, get_theme_icon(SNAME("ActionCopy"), SNAME("EditorIcons")), COPY_GROUP, false, TTR("Copy group name to clipboard."));
```

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
 
-			n = n->get_owner();
-		}
+	if (group_name.is_empty()) {
```
You could avoid repeating this code if you add a generic check method that takes ValidationPanel as an argument.

KoBeWi commented:
I think I'm done reviewing the code (in its current state) 🤔
The changes look fine overall, just need some polishing. I didn't test global renaming and removing, because it crashes currently.

Also not sure if it's related to this PR, but I noticed that the editor is much less responsive with launch and some Node operations. I'll test later.

DarkMessiah commented:
1) Can't reproduce this crash :'( 
2) I made the name column uneditable in the tree. Now when you double-click, a rename dialog will appear (same as in GroupsEditor): 
![image](https://github.com/godotengine/godot/assets/1756388/23fbcb36-7f64-4a23-9476-e0b6fddec39e)

3) Added `EditorProgress` to `rename_references()` and `remove_references()`.
4) Disabled "Rename" and "Delete" context menu options and shortcuts if the group can't be edited.
5) About the weird indentation: `CheckBox` cannot be disabled in `Tree`. Therefore, I used the `GuiCheckedDisabled` icon and `TreeItem::CELL_MODE_STRING` for these elements.

AThousandShips commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,556 @@
+}
+
+void GroupSettingsEditor::_check_rename() {
+	String group_name = rename_group->get_text().strip_edges();
```
```suggestion
	String new_name = rename_group->get_text().strip_edges();
```
For shadowing

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,556 @@
+		SceneTreeDock::get_singleton()->get_tree_editor()->update_tree();
+	}
+
+	for (RBSet<String>::Element *E = scenes.front(); E; E = E->next()) {
```
```suggestion
	for (const String &E : scenes) {
```

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,556 @@
+
+	for (RBSet<String>::Element *E = scenes.front(); E; E = E->next()) {
+		Ref<PackedScene> packed_scene = ResourceLoader::load(E->get());
+		rename.step(E->get(), step++);
+		if (packed_scene->get_state()->rename_group_references(StringName(p_old_name), StringName(p_new_name))) {
```
```suggestion
		rename.step(E->get(), step++);
		ERR_CONTINUE(packed_scene.is_null());
		if (packed_scene->get_state()->rename_group_references(StringName(p_old_name), StringName(p_new_name))) {
```
The crash happens when scene fails to load for some reason. This should fix it.

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,556 @@
+void GroupSettingsEditor::update_groups() {
+	if (updating_group) {
+		return;
+	}
+
+	updating_group = true;
+
+	groups_cache = ProjectSettings::get_singleton()->get_global_groups_list();
+
+	tree->clear();
+
+	TreeItem *root = tree->create_item();
```
Separating every line is a waste of space. I'd do it like this
```suggestion
	}
	updating_group = true;
	groups_cache = ProjectSettings::get_singleton()->get_global_groups_list();

	tree->clear();
	TreeItem *root = tree->create_item();
```

KoBeWi commented:
Global groups can still be edited in foreign nodes. They should be getting disabled like scene groups.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,791 @@ static bool can_edit(Node *p_node, String p_group) {
+			add->set_icon(get_theme_icon("Add", "EditorIcons"));
+		} break;
+		case NOTIFICATION_VISIBILITY_CHANGED: {
+			if (groups_dirty && is_visible_in_tree()) {
```
```suggestion
			if (groups_dirty && is_visible_in_tree()) {
				groups_dirty = false;
```
You reset it twice in the methods below. Could be done here, once.

KoBeWi commented on editor/groups_editor.cpp:
```diff
@@ -59,776 +61,791 @@ static bool can_edit(Node *p_node, String p_group) {
-	undo_redo->commit_action();
+void GroupsEditor::_global_group_changed() {
+	_update_groups();
+	call_deferred("_update_tree");
```
This can be changed to `callable_mp()` too (I mean here only; even if a method is bound, it's still better to call it directly when possible).

KoBeWi commented on editor/group_settings_editor.h:
```diff
@@ -0,0 +1,105 @@
+
+	HashMap<StringName, String> groups_cache;
+
+	bool updating_group = false;
```
```suggestion
	bool updating_groups = false;
```

akien-mga commented:
Needs rebase after #61818, you should now use the new `ProjectSettings::add_hidden_prefix` to hide the prefixed properties.

DarkMessiah commented:
Done!

DarkMessiah commented:
Rebased after https://github.com/godotengine/godot/commit/6de34fde27f650f738a1e46992f1b783f2cf9e76

reduz reviewed (dismissed):
This looks OK to me on the surface, but going through all scenes if you want to delete or rename a group seems very unfeasible to me. In a large project this could take an hour and become unusable.

I would try to think about other ways to do this, such as:
* Cache in EditorFileSystem the global groups used (that will probably be a good amount of work to do), so only the relevant scenes are changed.
* Find another way to do it, such as leaving the group remain there but consider it invalid on load if it does not exist (just ignore it). This seems probably like the easiest way.

reduz commented:
Maybe you could find some hybrid, such as implementing what you have already for renaming (and warn the user the whole project needs to be scanned to do this), while for deleting simply discard groups that don't exist.

Mickeon commented:
I would love the hybrid solution, simply because I would absolutely love to have this in the editor as soon as possible. It's been a long time coming, and the refinement after each maintainer's criticism is quite commendable in my opinion.

Further performance improvements regarding the implementation could be done at a later date, especially if they do **not** impact the user that isn't even making use of the groups feature extensively.

KoBeWi commented:
> Find another way to do it, such as leaving the group remain there but consider it invalid on load if it does not exist (just ignore it). This seems probably like the easiest way.

The problem is that if global group is removed, any leftovers will be considered local. Global and local groups are stored in the same way, they can be differentiated only based on the global group list. To make it work like you suggest we'd need to keep an explicit list of discarded groups (but that also means you can't re-use a removed group name).

akien-mga commented:
As we're reaching feature freeze and there seems to be some performance concerns for big projects, we decided in the production team to postpone to 4.3 and aim to merge it early on in that cycle (so likely in November) so we get enough time to iron out the quirks.

DarkMessiah commented:
I made a separate commit for caching. I'll do the squash after approval.

I used the same logic as for global classes and store all cache inside `scene_groups_cache.cfg`. I think the only problem is, if you change that cache file manually, you might lose all the cache information and renaming/deleting can never work correctly :(

```
[scene_path]
groups=[list of unique group names]
```

KoBeWi commented on scene/resources/packed_scene.cpp:
```diff
@@ -1849,6 +1849,44 @@ void SceneState::add_editable_instance(const NodePath &p_path) {
+
+HashSet<StringName> SceneState::get_all_groups() {
+	HashSet<StringName> ret;
+	for (int node_idx = 0; node_idx < get_node_count(); node_idx++) {
+		for (int group_idx = 0; group_idx < nodes[node_idx].groups.size(); group_idx++) {
+			ret.insert(names[nodes[node_idx].groups[group_idx]]);
+		}
+	}
```
```suggestion
	for (const NodeData &node : nodes) {
		for (const int &group : node.groups) {
			ret.insert(names[group]);
		}
	}
```
Prefer range iterators where possible (same in other methods).

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1634,6 +1651,65 @@ void EditorFileSystem::_queue_update_script_class(const String &p_path) {
+			continue;
+		}
+
+		HashSet<StringName> scene_groups = _get_scene_groups(path);
```
```suggestion
		const HashSet<StringName> scene_groups = _get_scene_groups(path);
```
The set is not modified, so it can be constant. Same for the method's return value.

AThousandShips commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,523 @@
+}
+
+void GroupSettingsEditor::_modify_references(const String &p_name, const String &p_new_name, bool p_is_rename) {
+	RBSet<String> scenes;
```
Any reason this is `RBSet` and not `HashSet`?

KoBeWi commented on editor/group_settings_editor.h:
```diff
@@ -72,8 +72,7 @@ class GroupSettingsEditor : public VBoxContainer {
 
-	void _get_node(const String &p_name);
-	void _get_all_scenes(EditorFileSystemDirectory *p_dir, RBSet<String> &r_list);
+	void _modify_references(const String &p_name, const String &p_new_name, bool p_is_rename);
```
Why does this take String arguments, not StringName?

KoBeWi commented:
I left some minor comments about the implementation, but the cache works great. Renaming all groups in a big project went down from a few minutes to a couple of seconds.

Though I noticed that when you rename or delete a scene, the old entry is kept in the cache.

DarkMessiah commented:
Done :) 

> Though I noticed that when you rename or delete a scene, the old entry is kept in the cache.

Connected to the `files_moved` and `file_removed` signals from FileSystemDock, now the cache is updating correctly.

>EDIT:
>Also a more serious issue is that modifying scene externally does not update cache entry. Not sure how to fix that; if the editor is currently opened then there is a way to detect that, but when it's not...
>It's very relevant when working in a team, as VCS causes lots of external changes. Though I think scripts have the same problem.

I don't know what to do about caching after external editing. I could suggest `Project -> Re-cache groups`, but this way seems to be the easiest but bad.

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,531 @@
+	group_name->set_h_size_flags(SIZE_EXPAND_FILL);
+	group_name->set_clear_button_enabled(true);
+	group_name->connect("text_changed", callable_mp(this, &GroupSettingsEditor::_group_name_text_changed));
+	group_name->connect("text_submitted", callable_mp(this, &GroupSettingsEditor::_add_button_pressed).unbind(1));
```
It might be worth it to bind to some intermediate method, because you can submit invalid name, triggering error dialog.

KoBeWi commented on scene/resources/packed_scene.cpp:
```diff
@@ -1849,6 +1849,44 @@ void SceneState::add_editable_instance(const NodePath &p_path) {
+	for (const NodeData &node : nodes) {
+		for (const int &group : node.groups) {
+			if (names[group] == p_old_name) {
+				names.set(group, p_new_name);
```
```suggestion
				names.write[group] = p_new_name;
```
Though `set()` is fine too I think (we just don't use it usually).

KoBeWi commented on editor/group_settings_editor.cpp:
```diff
@@ -0,0 +1,531 @@
+}
+
+void GroupSettingsEditor::remove_references(const StringName &p_name) {
+	_modify_references(p_name, "", false);
```
```suggestion
	_modify_references(p_name, StringName(), false);
```
Not sure how much it matters, might be optimized anyway.

KoBeWi reviewed (approved):
Seems to be working correctly now.

As for the external changes issue, scripts have the same problem and it's currently unresolved (AFAIK). Worst that it can cause is that renaming global group might miss some references, but I think renaming groups globally is not very common action.

DarkMessiah commented:
Removed hacks related to disabled Checkbox state, because Checkbox in Tree now has a disabled state: https://github.com/godotengine/godot/pull/84845

![image](https://github.com/godotengine/godot/assets/1756388/451680ee-d413-4950-884a-bd5fb89b6076)

YuriSizov commented:
The OP is unclear whether this closes https://github.com/godotengine/godot-proposals/issues/3351 or not. So, does it?

Exerionius commented:
Just saw this PR in the changelog for 4.3 dev1.

The description of local groups is kinda confusing.
> derived from nodes in the current scene

What does this mean?

Does this mean that if I have two separate scenes with a bunch of nodes in them, and both scenes define a group called "test", then calling `get_tree().get_nodes_in_group("test")` from one of the scenes will return nodes only from the same scene regardless if a group with the same name exists in a different scene?

Does this mean that the old behavior of groups was what is now called "global"? But now we have "local" groups in addition to that?

How can we know in code which group is local and which is global?

This needs to be thoroughly documented on this page: https://docs.godotengine.org/en/stable/tutorials/scripting/groups.html

Mickeon commented:
> The description of local groups is kinda confusing.
>
> > derived from nodes in the current scene
> 
> What does this mean?

> Does this mean that if I have two separate scenes with a bunch of nodes in them, and both scenes define a group called "test", then calling `get_tree().get_nodes_in_group("test")` from one of the scenes will return nodes only from the same scene regardless if a group with the same name exists in a different scene?

It's much simpler than you're imagining.

The underlying logic of groups is unchanged. Groups with the same name are the same.
The way groups are organized and displayed in the editor has changed drastically to make keeping track of them easier.

**All** groups are stored alongside the node they belong to in a **PackedScene**. When you load the scene, every node's group is kept track of and displayed in the Group Editor for easy access.
This is the exact same behavior as prior to this PR.

The only difference between "local" and "global" groups is...

> How can we know in code which group is local and which is global?

The "global" groups and their descriptions are stored in the Project Settings _(underneath `[global_group]` in the `.godot` file)_. The PR also implements some internal methods in the ProjectSettings class to retrieve them, but they are not exposed.

In code, you may write something like this _(I don't see why you'd need to but...)_:
```GDScript
static func is_global_group(group_name):
    return ProjectSetting.has_setting("global_group/" + group_name)

```

> This needs to be thoroughly documented on this page: https://docs.godotengine.org/en/stable/tutorials/scripting/groups.html

I wouldn't say "thoroughly" because again, your code would be entirely unaffected, but the page does need to be entirely refreshed for the new interface, yes.

Exerionius commented:
@Mickeon thanks for the explanations, it makes it a bit clearer.

So it's the opposite of what I was imagining: the old groups are what "local" means, but now we have "global" groups which is editable in the project settings and that's the only difference between the two (from the usage perspective).

fire commented:
Many people are able to reproduce a crash at https://github.com/godotengine/godot/issues/86844

aobrem commented:
The "Manage Groups" dialog provided a way to add multiple nodes to a group all at once. Is there still a way to do that after these changes?

Mickeon commented:
I'm afraid that's not as doable as before. That's mainly because the Node dock refuses to work when more than one Node is selected. I can see someone lifting this limitation in the future. It probably was done because it does not play nice with the Signals sub-section.

sancho2 commented:
>I can see someone lifting this limitation in the future.    
    
Do we need to do up a proposal for this or is it obvious to everyone that the manager is going to be missed?    
I just added 12 items to a group and the lack of a manage-window made even that small amount of nodes a small misery.  
This change is an incomplete change in the absence of an edit window.

KoBeWi commented:
It's best to open a proposal if it doesn't exist yet.

sancho2 commented:
[In my opinion this proposal covers it.](https://github.com/godotengine/godot-proposals/issues/10331)     
Thank you.

Mickeon commented:
It's a general quality of life. For anyone reading this with the know-how, they should feel free to open a PR. The implementation details should be discussed in the PR itself, I feel. Namely, what to do with the Signals tab on the side.
